File: 98. Validate Binary Search Tree/main.py
from typing import Optional

import numpy as np


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.helper(root, np.iinfo(np.int64).min, np.iinfo(np.int64).max)

    def helper(self, node: Optional[TreeNode], min: np.int64, max: np.int64) -> bool:
        if node is None:
            return True

        if node.val <= min or node.val >= max:
            return False

        return self.helper(node.left, min, node.val) and self.helper(
            node.right, node.val, max
        )


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.helper(root, float("-inf"), float("inf"))

    def helper(self, node: Optional[TreeNode], min: float, max: float) -> bool:
        if node is None:
            return True

        if node.val <= min or node.val >= max:
            return False

        return self.helper(node.left, min, node.val) and self.helper(
            node.right, node.val, max
        )


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        array = list()
        self.scan(array, root)
        for i in range(1, len(array)):
            if array[i - 1] >= array[i]:
                return False

        return True

    def scan(self, array: list, node: Optional[TreeNode]):
        if node is None:
            return

        self.scan(array, node.left)
        array.append(node.val)
        self.scan(array, node.right)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        array = list()
        self.scan(array, root)
        if array == sorted(array) and len(set(array)) == len(array):
            return True

        return False

    def scan(self, array: list, node: Optional[TreeNode]):
        if node is None:
            return

        self.scan(array, node.left)
        array.append(node.val)
        self.scan(array, node.right)

File: 98. Validate Binary Search Tree/main.cpp
#include <climits>
#include <cstdlib>
#include <vector>

using std::vector;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr) {}
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x)
        , left(left)
        , right(right) {}
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode *root) {
        return helper(root, LLONG_MIN, LLONG_MAX);
    }

private:
    bool helper(TreeNode *root, long long min, long long max) {
        if (root == nullptr) {
            return true;
        }

        if (root->val <= min || root->val >= max) {
            return false;
        }

        return helper(root->left, min, root->val) && helper(root->right, root->val, max);
    }
};

const size_t ARRAY_SIZE = 10000;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode *root) {
        TreeNode **array = (TreeNode **)calloc(ARRAY_SIZE, sizeof(TreeNode *));
        if (array == nullptr) {
            return false;
        }

        size_t i = 0;
        toArray(array, root, &i);
        for (i = 1; i < ARRAY_SIZE; i++) {
            if (array[i] == nullptr) {
                break;
            }

            if (array[i - 1]->val >= array[i]->val) {
                free(array);
                return false;
            }
        }

        free(array);
        return true;
    }

private:
    void toArray(TreeNode **array, TreeNode *node, size_t *index) {
        if (node == nullptr) {
            return;
        }

        toArray(array, node->left, index);
        array[*index] = node;
        *index += 1;
        toArray(array, node->right, index);
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode *root) {
        vector<int> v;

        scan(v, root);
        for (size_t i = 1; i < v.size(); i++) {
            if (v[i - 1] >= v[i]) {
                return false;
            }
        }

        return true;
    }

private:
    void scan(vector<int> &v, TreeNode *node) {
        if (node == nullptr) {
            return;
        }

        scan(v, node->left);
        v.push_back(node->val);
        scan(v, node->right);
    }
};